dm/pkg/shardddl/optimism/lock.go

```diff
@@ -548,7 +670,15 @@ func (l *Lock) addTables(tts []TargetTable) {
 			}
 			for table := range tables {
 				if _, ok := l.tables[tt.Source][schema][table]; !ok {
-					l.tables[tt.Source][schema][table] = l.joined
+					ti, err := l.FetchTableInfos(tt.Task, tt.Source, schema, table)
```

Comment:
As `FetchTableInfos` fetch table schema info from table checkpoint, I think it may cause fetch outdated table info since we do ddl before flush checkpoint. A example:
Worker1:  --ddl1(tbl1) --------  execDDL1   -------- flushTbl1CP ------------->
Worker2:  --------- ddl2(tbl2) ------------- execDDL2 ----- flushTbl2Cp--->
Master:      -----------handleDDL1 --- handleDDL2------------------------->

In this case, when master start handle DDL2 (tbl2), tbl1's checkpoint didn't finished flushing, thus master will fetch outdated table schema for tbl1 and it may impact the correctness of the result.

Reply:
This is a problem, but I don't think it's easy to solve under the existing architecture and it's not a big problem. First, `fetchtableinfo` is only called when the master receives the first DDL (to create a lock), that is, in most cases, DDL2 will not cause the master to call `fetchtableinfo`. In addition, we do not guarantee that DDLs under different sources are executed in order, that is to say, it is uncertain who executes DDL1 and DDL2 first, so the master can only coordinate according to the existing information. For source1, DDL1's Coordination is fine, for source2 there will be a problem if and only if DDL1 and DDL2 are updating the same column (eg DDL1 add column a, DDL2 drop column a). Finally, unless we can ensure that executing DDL and flushing checkpoint are atomic, this problem is difficult to solve. We have had such problems before, such as worker1 crash without flushing checkpoint after executing DDL1, worker2 executing DDL2, and then when master coordinating DDL2, it use the table structure after worker1 executes DDL1.

Reply:
LGTM.